File: 394__Decode_String.py
class Solution(object):
    def decodeString(self, s):
        """
        :type s: str
        :rtype: str
        """
        stack = []
        n = 0
        for char in s:
            if char.isdigit():
                n = n * 10 + int(char)
            elif char == "[":
                stack.append(str(n))
                n = 0
            elif char == "]":
                tmp = ""
                while stack and stack[-1].isalpha():
                    tmp = stack.pop() + tmp
                if stack:
                    num = stack.pop()
                stack.append(int(num) * tmp)
            else:
                stack.append(char)

        return "".join(stack)

File: Java/394__Decode_String.java
class Solution {
    public String decodeString(String s) {
        Stack<Integer> nums = new Stack<>();
        Stack<String> strs = new Stack<>();
        String result = "";
        int i = 0;
        while (i < s.length()) {
            char c = s.charAt(i);
            if (Character.isDigit(c)) {
                int val = 0;
                while (Character.isDigit(s.charAt(i))) {
                    val = val * 10 + (s.charAt(i) - '0');
                    i++;
                }
                nums.push(val);
            } else if (c == '[') {
                strs.push(result);
                result = "";
                i++;
            } else if (c == ']') {
                String last_str = strs.pop();
                int last_val = nums.pop();
                StringBuilder temp = new StringBuilder(last_str);
                for (int j = 0; j < last_val; j++){
                    temp.append(result);
                }
                result = temp.toString();
                i++;
            } else {
                result += c;
                i++;
            }
        }

        return result;
    }
}

    
